core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionImpl.java

```diff
@@ -78,23 +88,67 @@
                 parameters[i] = ctx;
             } else if (entry.getType().isAssignableFrom(HttpRequest.class)) {
                 parameters[i] = req;
+            } else if (entry.getType().isAssignableFrom(AggregatedHttpMessage.class)) {
+                parameters[i] = message;
             }
         }
         return parameters;
     }
 
     @Override
     public Object serve(ServiceRequestContext ctx, HttpRequest req, Map<String, String> args) throws Exception {
-        Object[] parameters = parameters(ctx, req, args);
+        if (aggregationRequired) {
+            if (CompletionStage.class.isAssignableFrom(method.getReturnType())) {
+                return req.aggregate()
+                          .thenComposeAsync(msg -> (CompletionStage<?>) invokeMethod(ctx, req, args, msg),
+                                            ctx.contextAwareEventLoop());
+            } else if (isAsynchronous) {
+                return req.aggregate()
+                          .thenApplyAsync(msg -> invokeMethod(ctx, req, args, msg),
+                                          ctx.contextAwareEventLoop());
```

Comment:
It was fixed by this comment: https://github.com/line/armeria/pull/567#pullrequestreview-40441811

It's to make RequestContext is set into thread local. If I used `thenApply` and `thenCompose`, I got a `IllegalStateException` when I called `RequestContext.current()` in an annotated method. Because my first implementation didn't include wrapping the existing RequestContext instance and didn't push the context into thread local.

But, I think if I called `RequestContext.push(ctx, false)` before invoking a method, `thenApply` and `thenCompose` would be OK. Is this better design? WDYT?